### PR TITLE
[GOBBLIN-2018] Add Opentelemetry metrics for GaaS observability

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2018

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

    Gobblin-as-a-Service uses Kafka events to send additional metadata from flows to enable user analytics on their flows. We want to also support opentelemetry as a method to visualize and alert based on flow statuses.
    
    `metrics.reporting.opentelemtry.metrics.enabled=true` to enable opentelemetry metrics and `metrics.reporting.opentelemtry.configs` to send specific configurations to the opentelemetry reporter
    

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Brought in in-memory Otel libraries and integrated them to assert metric values and their associated tags. Future OpenTelemetry metric tests should utilize `InMemoryOpentelemetryMetrics` to get a consistent way to observe metrics collected by their reporters.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/InMemoryOpenTelemetryMetrics.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * A stub for OpenTelemetryMetrics that uses in-memory metric reader for testing purposes
+ */
+public class InMemoryOpenTelemetryMetrics extends OpenTelemetryMetricsBase {
+  public InMemoryMetricReader metricReader;
+  public InMemoryOpenTelemetryMetrics(State state) {
+    super(state);
+  }
+  @Override
```
Comment (arjun4084346) [inline gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/InMemoryOpenTelemetryMetrics.java:36]: blank line here

File: gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Override
+  protected MetricExporter initializeMetricExporter(State state) {
+    OtlpHttpMetricExporterBuilder httpExporterBuilder = OtlpHttpMetricExporter.builder();
+    httpExporterBuilder.setEndpoint(state.getProp(ConfigurationKeys.METRICS_REPORTING_OPENTELEMETRY_ENDPOINT));
+    return httpExporterBuilder.build();
+  }
+
+  public static OpenTelemetryMetrics getInstance(State state) {
+    if (state.getPropAsBoolean(ConfigurationKeys.METRICS_REPORTING_OPENTELEMETRY_ENABLED, ConfigurationKeys.DEFAULT_METRICS_REPORTING_OPENTELEMETRY_ENABLED)) {
```
Comment (arjun4084346) [inline gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java:60]: we can combine these two conditions right?

File: gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.export.MetricExporter;
+import io.opentelemetry.sdk.metrics.export.PeriodicMetricReader;
+import io.opentelemetry.sdk.resources.Resource;
+
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.util.PropertiesUtils;
+/**
```
Comment (arjun4084346) [inline gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java:39]: blank line

File: gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ * A metrics reporter wrapper that uses the OpenTelemetry standard to emit metrics
+ * Currently separated from the legacy codehale metrics as we need to maintain backwards compatibility, but eventually
+ * can replace the old metrics system with tighter integrations once it's stable
+ */
+public class OpenTelemetryMetrics extends OpenTelemetryMetricsBase {
+
+  private static OpenTelemetryMetrics GLOBAL_INSTANCE;
+  private static final Long DEFAULT_OPENTELEMETRY_REPORTING_INTERVAL_MILLIS = 10000L;
+  private OpenTelemetryMetrics(State state) {
```
Comment (arjun4084346) [inline gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/OpenTelemetryMetrics.java:48]: blank line

**Review by arjun4084346 — APPROVED** [review]

**Review by umustafi — APPROVED** [review]
How are you testing the in memory metrics if they're not reported anywhere?

**Comment by Will-Lo** [comment]
Opentelemetry decouples metrics emission from the way it reads from your metrics. The metric reader specification is here: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.13.0/specification/metrics/sdk.md#metricreader

So essentially you can have pluggable exporters that work with any reader, you can have the reader follow some aggregation and other behaviors e.g. scheduled, and exporters only deal with sending metrics to a backend system over a communication mechanism.

The purpose of using the in-memory metric reader is to read directly from the metrics instead of having to wait for the scheduled reader that is the default behavior. This helps with avoiding race conditions in tests although care needs to be taken with the aggregation settings (left as default similar to the scheduled reader for now).

**Pull request merged**